File: backend/api/routes/notifications.py
from typing import Any

from fastapi import APIRouter, HTTPException, Query, Request
from pydantic import BaseModel, Field

router = APIRouter()


class NotificationCreatePayload(BaseModel):
    id: str | None = Field(default=None, max_length=100)
    type: str = Field(..., min_length=1, max_length=50)  # 'sediment', 'glitch', 'trace'
    snippet: str = Field(..., min_length=1, max_length=10_000)
    timestamp: str | None = Field(default=None, max_length=50)
    conversation_id: str | None = Field(default=None, max_length=100)
    message_id: int | None = None
    parent_message_id: int | None = None
    speaker: str | None = Field(default=None, max_length=100)
    source: str | None = Field(default=None, max_length=500)
    read: int = 0
    dismissed: int = 0
    source_type: str | None = Field(default=None, max_length=100)
    source_id: str | None = Field(default=None, max_length=100)


class ClearPayload(BaseModel):
    type: str | None = Field(default=None, max_length=50)


@router.get("/notifications", response_model=list[dict[str, Any]])
def list_notifications(
    dismissed: bool | None = None,
    type: str | None = Query(default=None, max_length=50),
    search: str | None = Query(default=None, max_length=500),
    limit: int = Query(default=50, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
    request: Request = None,
):
    state = request.app.state
    notification_repo = state.notification_repo
    return notification_repo.list_all(
        limit=limit,
        offset=offset,
        dismissed=dismissed,
        type_filter=type,
        search_query=search,
    )


@router.get("/notifications/{id}", response_model=dict[str, Any])
def get_notification(id: str, request: Request = None):
    state = request.app.state
    notification_repo = state.notification_repo
    notif = notification_repo.get(id)
    if not notif:
        raise HTTPException(status_code=404, detail="Notification not found")
    return notif


@router.post("/notifications", response_model=dict[str, Any])
def create_notification(
    payload: NotificationCreatePayload,
    request: Request = None,
):
    state = request.app.state
    notification_repo = state.notification_repo
    try:
        return notification_repo.create(
            type=payload.type,
            snippet=payload.snippet,
            id=payload.id,
            timestamp=payload.timestamp,
            conversation_id=payload.conversation_id,
            message_id=payload.message_id,
            parent_message_id=payload.parent_message_id,
            speaker=payload.speaker,
            source=payload.source,
            read=payload.read,
            dismissed=payload.dismissed,
            source_type=payload.source_type,
            source_id=payload.source_id,
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e)) from e


@router.patch("/notifications/{id}/read", response_model=dict[str, Any])
def mark_read(id: str, request: Request = None):
    state = request.app.state
    notification_repo = state.notification_repo
    notif = notification_repo.mark_as_read(id)
    if not notif:
        raise HTTPException(status_code=404, detail="Notification not found")
    return notif


@router.patch("/notifications/{id}/unread", response_model=dict[str, Any])
def mark_unread(id: str, request: Request = None):
    state = request.app.state
    notification_repo = state.notification_repo
    notif = notification_repo.mark_as_unread(id)
    if not notif:
        raise HTTPException(status_code=404, detail="Notification not found")
    return notif


@router.patch("/notifications/{id}/dismiss", response_model=dict[str, Any])
def dismiss_notification(id: str, request: Request = None):
    state = request.app.state
    notification_repo = state.notification_repo
    notif = notification_repo.dismiss(id)
    if not notif:
        raise HTTPException(status_code=404, detail="Notification not found")
    return notif


@router.patch("/notifications/dismiss-match", response_model=dict[str, Any])
def dismiss_by_match_endpoint(
    conversation_id: str,
    message_id: int,
    request: Request = None,
) -> dict[str, Any]:
    state = request.app.state
    notification_repo = state.notification_repo
    notification_repo.dismiss_by_match(conversation_id, message_id)
    return {"status": "ok"}


@router.post("/notifications/clear")
def clear_notifications(
    payload: ClearPayload | None = None,
    request: Request = None,
):
    state = request.app.state
    notification_repo = state.notification_repo
    if payload and payload.type:
        notification_repo.clear_by_type(payload.type)
    else:
        notification_repo.clear_all()
    return {"status": "ok"}


@router.post("/notifications/read")
def mark_all_read_endpoint(
    payload: ClearPayload | None = None,
    request: Request = None,
):
    state = request.app.state
    notification_repo = state.notification_repo
    if payload and payload.type:
        notification_repo.mark_all_as_read(payload.type)
    else:
        notification_repo.mark_all_as_read()
    return {"status": "ok"}

File: frontend/src/api/notifications.ts
import { BASE } from "./http"
import type { SedimentNotification } from "./types"

export { type SedimentNotification }

function mapBackendNotification(n: any): SedimentNotification {
  return { id: n.id, type: n.type, conversationId: n.conversation_id || undefined, messageId: n.message_id || undefined, parentMessageId: n.parent_message_id || undefined, timestamp: n.timestamp, snippet: n.snippet, speaker: n.speaker || undefined, source: n.source || undefined, sourceType: n.source_type || undefined, sourceId: n.source_id || undefined, read: n.read === 1 || n.read === true, dismissed: n.dismissed === 1 || n.dismissed === true }
}

function mapFrontendNotification(n: Partial<SedimentNotification>): any {
  return { id: n.id, type: n.type, conversation_id: n.conversationId, message_id: n.messageId, parent_message_id: n.parentMessageId, timestamp: n.timestamp, snippet: n.snippet, speaker: n.speaker, source: n.source, source_type: n.sourceType, source_id: n.sourceId, read: n.read ? 1 : 0, dismissed: n.dismissed ? 1 : 0 }
}

export async function getNotification(id: string): Promise<SedimentNotification> {
  const res = await fetch(`${BASE}/notifications/${id}`)
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return mapBackendNotification(await res.json())
}

export async function getNotifications(dismissed?: boolean, limit = 100, type?: string, search?: string): Promise<SedimentNotification[]> {
  const params = new URLSearchParams({ limit: String(limit) })
  if (dismissed !== undefined) params.set("dismissed", dismissed ? "true" : "false")
  if (type) params.set("type", type)
  if (search) params.set("search", search)
  const res = await fetch(`${BASE}/notifications?${params}`)
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return (await res.json()).map(mapBackendNotification)
}

export async function createNotification(notif: Partial<SedimentNotification>): Promise<SedimentNotification> {
  const res = await fetch(`${BASE}/notifications`, { method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify(mapFrontendNotification(notif)) })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return mapBackendNotification(await res.json())
}

export async function markNotificationRead(id: string): Promise<SedimentNotification> {
  const res = await fetch(`${BASE}/notifications/${id}/read`, { method: "PATCH" })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return mapBackendNotification(await res.json())
}

export async function markNotificationUnread(id: string): Promise<SedimentNotification> {
  const res = await fetch(`${BASE}/notifications/${id}/unread`, { method: "PATCH" })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return mapBackendNotification(await res.json())
}

export async function dismissNotification(id: string): Promise<SedimentNotification> {
  const res = await fetch(`${BASE}/notifications/${id}/dismiss`, { method: "PATCH" })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
  return mapBackendNotification(await res.json())
}

export async function dismissNotificationByMatch(conversationId: string, messageId: number): Promise<void> {
  const params = new URLSearchParams({ conversation_id: conversationId, message_id: String(messageId) })
  const res = await fetch(`${BASE}/notifications/dismiss-match?${params}`, { method: "PATCH" })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
}

export async function clearNotifications(type?: string): Promise<void> {
  const res = await fetch(`${BASE}/notifications/clear`, { method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify({ type }) })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
}

export async function markAllNotificationsRead(type?: string): Promise<void> {
  const res = await fetch(`${BASE}/notifications/read`, { method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify({ type }) })
  if (!res.ok) throw new Error(`HTTP ${res.status}`)
}
